infrastructure/assumptions/non-local-ports.sub.window.js

```diff
@@ -30,38 +37,71 @@ function resolveUrl(url, options) {
   return result;
 }
 
-promise_test(async () => {
-  const url = resolveUrl("/common/blank-with-cors.html", {
+const kOptions = {
+  httpPrivate: {
+    protocol: "http:",
+    port: kPorts.httpPrivate,
+  },
+  httpsPrivate: {
     protocol: "https:",
-    port: "{{ports[https-private][0]}}"
-  });
+    port: kPorts.httpsPrivate,
+  },
+  httpPublic: {
+    protocol: "http:",
+    port: kPorts.httpPublic,
+  },
+  httpsPublic: {
+    protocol: "https:",
+    port: kPorts.httpsPublic,
+  },
+};
+
+promise_test(async () => {
+  const url = resolveUrl("/common/blank-with-cors.html", kOptions.httpsPrivate);
   const response = await fetch(url);
   assert_true(response.ok);
 }, "Fetch from https-private port works.");
 
 promise_test(async () => {
-  const url = resolveUrl("/common/blank-with-cors.html", {
-    protocol: "http:",
-    port: "{{ports[http-private][0]}}"
-  });
+  const url = resolveUrl("/common/blank-with-cors.html", kOptions.httpPrivate);
   const response = await fetch(url);
   assert_true(response.ok);
 }, "Fetch from http-private port works.");
 
 promise_test(async () => {
-  const url = resolveUrl("/common/blank-with-cors.html", {
-    protocol: "https:",
-    port: "{{ports[https-public][0]}}"
-  });
+  const url = resolveUrl("/common/blank-with-cors.html", kOptions.httpsPublic);
   const response = await fetch(url);
   assert_true(response.ok);
 }, "Fetch from https-public port works.");
 
 promise_test(async () => {
-  const url = resolveUrl("/common/blank-with-cors.html", {
-    protocol: "http:",
-    port: "{{ports[http-public][0]}}"
-  });
+  const url = resolveUrl("/common/blank-with-cors.html", kOptions.httpPublic);
   const response = await fetch(url);
   assert_true(response.ok);
 }, "Fetch from http-public port works.");
+
+promise_test(async (t) => {
+  const futureMessage = new Promise((resolve) => {
+    window.addEventListener("message", resolve);
+  });
+
+  const iframe = await new Promise((resolve, reject) => {
+    const iframe = document.createElement("iframe");
+    iframe.src =
+      resolveUrl("resources/fetch-and-post-result.html", kOptions.httpPublic);
+
+    iframe.onload = () => { resolve(iframe); };
+    iframe.onerror = reject;
+
+    document.body.appendChild(iframe);
+    t.add_cleanup(() => {
+      document.body.removeChild(iframe);
+    });
+  });
+
+  iframe.contentWindow.postMessage(
+    resolveUrl("/common/blank-with-cors.html").toString(), "*");
+
+  const evt = await futureMessage;
+  assert_equals(evt.data, "TypeError: Failed to fetch");
```

Comment:
This depends on the error message. https://wpt.fyi/results/infrastructure/assumptions/non-local-ports.sub.window.html?label=pr_head&max-count=1&pr=29248 shows Firefox and Safari failing because the fetch succeeded, but if they implement this they'll most likely fail due to not having exactly "Failed to fetch" as the error message.

I'll suggest below what to do instead.

Reply:
Makes sense. See below.